src/goto-instrument/expand_pointer_predicates.cpp

```diff
@@ -0,0 +1,211 @@
+      simplify(size_expr, ns);
+
+      // This should be forced by typechecking.
+      INVARIANT(
+        pointer_expr.type().id() == ID_pointer && is_lvalue(pointer_expr),
+        "Invalid argument to valid_pointer.");
+
+      local_bitvector_analysist::flagst flags =
+        local_bitvector_analysis.get(target, pointer_expr);
+      // Only create a new object if the pointer may be uninitialized.
+      if(flags.is_uninitialized() || flags.is_unknown())
+      {
+        typet object_type = type_from_size(size_expr, ns);
+
+        // Decl a new variable (which is therefore unconstrained)
+        goto_programt::targett d = assume_code.add_instruction(DECL);
+        d->source_location = assume_expr.source_location();
+        const symbol_exprt obj =
+          new_tmp_symbol(object_type, d->source_location).symbol_expr();
+        d->code = code_declt(obj);
+
+        exprt rhs;
+        if(object_type.id() == ID_array)
+        {
+          rhs = typecast_exprt(
+            address_of_exprt(index_exprt(obj, from_integer(0, index_type()))),
+            pointer_expr.type());
+        }
+        else
+        {
+          rhs = address_of_exprt(obj);
+        }
+
+        // Point the relevant pointer to the new object
+        goto_programt::targett a = assume_code.add_instruction(ASSIGN);
+        a->source_location = assume_expr.source_location();
+        a->code = code_assignt(pointer_expr, rhs);
+        a->code.add_source_location() = assume_expr.source_location();
+      }
+
+      // Because some uses of this occur before deallocated, dead, and malloc
+      // objects are initialized, we need to make some additional assumptions
+      // to clarify that our newly allocated object is not dead, deallocated,
+      // or outside the bounds of a malloc region.
+
+      exprt check_expr =
+        points_to_valid_memory_def(pointer_expr, size_expr, ns);
+      valid_memory_expr.swap(check_expr);
+      it.next_sibling_or_parent();
+    }
+    else
+    {
+      ++it;
+    }
+  }
+
+  program.destructive_insert(target, assume_code);
+}
+
+void expand_pointer_predicatest::expand_assertion(exprt &assert_expr)
+{
+  for(depth_iteratort it = assert_expr.depth_begin();
+      it != assert_expr.depth_end();)
+  {
+    if(it->id() == ID_points_to_valid_memory)
+    {
+      // Build an expression that checks validity.
+      exprt &valid_memory_expr = it.mutate();
+      exprt &pointer_expr = valid_memory_expr.op0();
+      exprt &size_expr = valid_memory_expr.op1();
+
+      exprt check_expr =
+        points_to_valid_memory_def(pointer_expr, size_expr, ns);
+      valid_memory_expr.swap(check_expr);
+      it.next_sibling_or_parent();
+    }
+    else
+    {
+      ++it;
+    }
+  }
+}
+
+const symbolt &expand_pointer_predicatest::new_tmp_symbol(
+  const typet &type,
+  const source_locationt &source_location)
+{
+  return get_fresh_aux_symbol(
+    type,
+    id2string(source_location.get_function()),
```

Comment:
If possible, please use the actual function identifier as found in the function map instead of this user-level name. This will require passing in such an additional `irep_idt` to `new_tmp_symbol`.